src/hooks/use-file-upload-reducer.tsx

```diff
@@ -0,0 +1,84 @@
+import * as React from 'react'
+import find from 'lodash/find'
+
+// TODO: add `id` for finding instead of `file`
+type FileUpload = {
+  file: File
+  percent: number
+  message: string
+  signedUrl?: string
+}
+
+type FileUploadReducerState = {files: FileUpload[]}
+
+type Action = {
+  type: 'add' | 'progress' | 'finalize'
+}
+
+type Actions = {
+  add: {
+    fileUpload: FileUpload
+  }
+  progress: FileUpload
+  finalize: {
+    file: FileUpload['file']
+    fileUrl: string
+  }
+}
+
+type ReducerAction = {
+  [ActionType in keyof Actions]: {
+    type: ActionType
+  } & Actions[ActionType]
+}[keyof Actions]
+
+const fileUploadReducer = (
```

Comment:
did y'all consider xstate here? just curious, it likely works out very similarly but xstate has a wider safety net. 

not a call for change!